### PR TITLE
Fix issues with A* expansion

### PR DESCRIPTION
**Pull request by smWork**
1. End when we surpass a certain cost (instead of # nodes in queue).
2. Enqueue next nodes immediately after dequeuing current node.

### CODE REVIEW

**Comment by smWork** [comment]
@dnesbitt61 @kevinkreiser I sent an email about this change. Please review.

**Comment by dnesbitt61** [comment]
Are you sure this isn't a data import issue? I thought the logic was generally sound: do not allow expansion to enter a destination only region once you exit the destination only region - figuring the other direction would exit the region and a connection would be made.

I prefer a multiple pass approach I think - if the first pass fails, run again with the destination only logic disabled as this PR does. Would something like that work?

**Comment by smWork** [comment]
Not a data import issue. Real world data usually contains some destination-only donuts. But, I like your approach to disable destination-only in first pass and allow it with penalty if first pass fails. Will make that change shortly.

Majority of my changes are related to the A* expansion itself. The issues it addresses will apply to any large transition cost (country boundaries, ferry etc). Please review that and let me know if you find issues with:
1. End when we surpass a certain cost (instead of # nodes in queue)
2. Enqueue next nodes immediately after dequeuing current node

**Comment by smWork** [comment]
I will update this PR with proposed changes. Please do not merge this yet.

File: src/sif/autocost.cc [hunk 1]
```diff
@@ -76,13 +76,6 @@ class AutoCost : public DynamicCost {
   virtual bool AllowMultiPass() const;
 
   /**
-   * Disables entrance into destination only areas. This should only be used
```
Comment (smWork) [inline src/sif/autocost.cc]: Removed override of DisableDestinationOnly. The base-class function is sufficient.

File: src/sif/autocost.cc [hunk 2]
```diff
@@ -408,7 +398,7 @@ Cost AutoCost::TransitionCost(const baldr::DirectedEdge* edge,
 
   // Additional penalties without any time cost
   uint32_t idx = pred.opp_local_idx();
-  if (!pred.destonly() && edge->destonly()) {
+  if (!disable_destination_only_ && !pred.destonly() && edge->destonly()) {
```
Comment (smWork) [inline src/sif/autocost.cc]: Do not penalize destination-only if it's disabled.

File: src/thor/bidirectional_astar.cc [hunk 3]
```diff
@@ -369,11 +365,42 @@ std::vector<PathInfo> BidirectionalAStar::GetBestPath(PathLocation& origin,
       forward_pred_idx = adjacencylist_forward_->pop();
       if (forward_pred_idx != kInvalidLabel) {
         // Check if the edge on the forward search connects to a
-        // reached edge on the reverse search tree.
+        // settled edge on the reverse search tree.
         pred = edgelabels_forward_[forward_pred_idx];
-        if (edgestatus_reverse_->Get(pred.opp_edgeid()).set() != EdgeSet::kUnreached) {
+        if (edgestatus_reverse_->Get(pred.opp_edgeid()).set() == EdgeSet::kPermanent) {
```
Comment (smWork) [inline src/thor/bidirectional_astar.cc]: Consider valid path only when it's settled from the other end. Makes sure transition cost is taken into account.

File: src/thor/route_action.cc [hunk 4]
```diff
@@ -94,6 +94,7 @@ namespace valhalla {
         float relax_factor = using_astar ? 16.0f : 8.0f;
         float expansion_within_factor = using_astar ? 4.0f : 2.0f;
         cost->RelaxHierarchyLimits(relax_factor, expansion_within_factor);
+        cost->EnableDestinationOnly();
```
Comment (smWork) [inline src/thor/route_action.cc]: If route fails, re-try with destination-only enabled.

**Comment by dnesbitt61** [comment]
I am confused. My thought was to run the first pass with Destination Only logic (for performance reasons) and to disable destination only on the second pass if the first pass fails.

File: src/thor/bidirectional_astar.cc (hunk 3 above)
Comment (dnesbitt61) [inline src/thor/bidirectional_astar.cc]: This makes sense to me. I forget why I allowed marking initial connections that were not settled. Testing further to see what impact (e.g. route changes/performance) this has.

**Comment by smWork** [comment]
This PR is no longer needed. @dnesbitt61 incorporated the changes in https://github.com/valhalla/valhalla/pull/663.

**Pull request closed without merging**